js/test/integration-tests.ts

```diff
@@ -0,0 +1,114 @@
+        for (let i = -1, n = props.length; ++i < n;) {
+            const prop = props[i];
+            if (this.utils.stringify(v1[prop]) !== this.utils.stringify(v2[prop])) {
+                propsFailures.push(`${prop}: ${format(v1[prop], v2[prop], ' !== ')}`);
+            }
+        }
+
+        for (let i = -1, n = v1.length; ++i < n;) {
+            let x1 = v1.get(i), x2 = v2.get(i);
+            if (this.utils.stringify(x1) !== this.utils.stringify(x2)) {
+                getFailures.push(`${i}: ${format(x1, x2, ' !== ')}`);
+            }
+        }
+
+        let i = -1;
+        for (let [x1, x2] of zip(v1, v2)) {
+            ++i;
+            if (this.utils.stringify(x1) !== this.utils.stringify(x2)) {
+                iteratorFailures.push(`${i}: ${format(x1, x2, ' !== ')}`);
+            }
+        }
+
+        return {
+            pass: allFailures.every(({ failures }) => failures.length === 0),
+            message: () => [
+                `${v1.name}: (${format('cpp', 'java', ' !== ')})\n`,
+                ...allFailures.map(({ failures, title }) =>
+                    !failures.length ? `` : [`${title}:`, ...failures].join(`\n`))
+            ].join('\n')
+        };
+    }
+});
+
+describe(`Integration`, () => {
+    for (const format of formats) {
+        describe(format, () => {
+            for (const [cppArrow, javaArrow] of zip(config.cpp[format], config.java[format])) {
+                describe(`${cppArrow.name}`, () => {
+                    testReaderIntegration(cppArrow.buffers, javaArrow.buffers);
+                    testTableFromBuffersIntegration(cppArrow.buffers, javaArrow.buffers);
+                });
+            }
+        });
+    }
+});
+
+function testReaderIntegration(cppBuffers: Uint8Array[], javaBuffers: Uint8Array[]) {
+    test(`cpp and java vectors report the same values`, () => {
+        expect.hasAssertions();
+        for (const [cppVectors, javaVectors] of zip(readVectors(cppBuffers), readVectors(javaBuffers))) {
+            expect(cppVectors.length).toEqual(javaVectors.length);
+            for (let i = -1, n = cppVectors.length; ++i < n;) {
+                (expect(cppVectors[i]) as any).toEqualVector(javaVectors[i]);
+            }
+        }
+    });
+}
+
+function testTableFromBuffersIntegration(cppBuffers: Uint8Array[], javaBuffers: Uint8Array[]) {
+    test(`cpp and java tables report the same values`, () => {
+        expect.hasAssertions();
+        const cppTable = Table.from(cppBuffers);
+        const javaTable = Table.from(javaBuffers);
+        const cppVectors = cppTable.columns;
+        const javaVectors = javaTable.columns;
+        expect(cppTable.length).toEqual(javaTable.length);
+        expect(cppVectors.length).toEqual(javaVectors.length);
+        for (let i = -1, n = cppVectors.length; ++i < n;) {
+            (expect(cppVectors[i]) as any).toEqualVector(javaVectors[i]);
+        }
+    });
+}
```

Comment:
I'm a bit confused by these integration tests, maybe I'm missing something. Shouldn't JavaScript be reading the JSON and comparing what it thinks are the contents of the JSON with the binary coming from Java and C++? Here it looks like it is reading the C++ and Java binary files and comparing them.

Reply:
@wesm yes they should. @TheNeuralBit is right, the goal here was to get the JS version reading C++ and Java arrows identically, and this test is how I validate that it does. Right now it _doesn't_ validate that it reads correctly (that's what the other tests do), just that there's no difference between the two. I think the plan is to update this test when we do have the JSON reader/writer in JS